[5.4.1] build failure on x32

5.3.0 built fine, so probably something to do with the recent avx changes in 5.4.x. @animetosho ?

```
x86_64-linux-gnux32-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -specs=/usr/share/dpkg/pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DCRCUTIL_USE_ASM=0 -I/usr/include/python3.10 -c src/yencode/encoder_vbmi2.cc -o /<<PKGBUILDDIR>>/.pybuild/cpython3_3.10_sabyenc3/src/yencode/encoder_vbmi2.o -Wall -Wextra -Wno-unused-function -fomit-frame-pointer -fno-rtti -fno-exceptions -O3 -fPIC -fwrapv -std=c++11 -mavx512vbmi2 -mavx512vl -mavx512bw -mpopcnt -mbmi -mbmi2 -mlzcnt
src/yencode/encoder_avx_base.h: Assembler messages:
src/yencode/encoder_avx_base.h:296: Error: incorrect register `%eax' used with `q' suffix
error: command '/usr/bin/x86_64-linux-gnux32-gcc' failed with exit code 1
```

Full build logs here: [x32](https://buildd.debian.org/status/fetch.php?pkg=python-sabyenc&arch=x32&ver=5.4.1-1&stamp=1653385037&raw=0) | [all](https://buildd.debian.org/status/package.php?p=python-sabyenc&suite=experimental)
Toolchain package versions: g++-11_11.2.0-14 gcc-11_11.2.0-14 libc6-dev_2.33-7 libstdc++-11-dev_11.2.0-14 libstdc++6_12.1.0-2 linux-libc-dev_5.17.6-1+b1

--- a/src/yencode/decoder_avx2_base.h
+++ b/src/yencode/decoder_avx2_base.h
@@ -64,10 +64,12 @@ HEDLEY_ALWAYS_INLINE void do_decode_avx2(const uint8_t* HEDLEY_RESTRICT src, lon
     // for some reason, MSVC Win32 seems to crash when trying to compile _mm256_mask_cmpeq_epi8_mask
     // the crash can be fixed by switching the order of the last two arguments, but it seems to generate wrong code
     // so just disable the optimisation as it seems to be problematic there
-#if defined(_MSC_VER) && !defined(PLATFORM_AMD64) && !defined(__clang__)
+#if defined(__AVX512VL__) && defined(__AVX512BW__)
+# if defined(_MSC_VER) && !defined(PLATFORM_AMD64) && !defined(__clang__)
     const bool useAVX3MaskCmp = false;
-#else
+# else
     const bool useAVX3MaskCmp = (use_isa >= ISA_LEVEL_AVX3);
+# endif
 #endif
     intptr_t i;
     for(i = -len; i; i += sizeof(__m256i)*2) {

--- a/src/yencode/decoder_neon.cc
+++ b/src/yencode/decoder_neon.cc
@@ -20,8 +20,8 @@
 #endif
 
 
-// for compilers that lack these functions
-#if defined(__clang__) || (defined(__GNUC__) && (defined(__aarch64__) && __GNUC__ >= 8))
+// for compilers that lack these functions (Clang armv7 9-12 seems to have issues with multi-vector loads)
+#if (defined(__clang__) && (defined(__aarch64__) || __clang_major__<9 || __clang_major__>12)) || (defined(__GNUC__) && (defined(__aarch64__) && __GNUC__ >= 8))
 # define vld1q_u8_x2_align(p, n) vld1q_u8_x2((uint8_t*)__builtin_assume_aligned(p, n))
 #else
 # define vld1q_u8_x2_align(p, n) vcreate2_u8(vld1q_u8_align(p, (n)/2), vld1q_u8_align((p)+16, (n)/2))

--- a/src/yencode/decoder_sse_base.h
+++ b/src/yencode/decoder_sse_base.h
@@ -122,10 +122,12 @@ HEDLEY_ALWAYS_INLINE void do_decode_sse(const uint8_t* HEDLEY_RESTRICT src, long
 #else
     const bool _USING_BLEND_ADD = false;
 #endif
-#if defined(_MSC_VER) && !defined(PLATFORM_AMD64) && !defined(__clang__)
+#if defined(__AVX512VL__) && defined(__AVX512BW__)
+# if defined(_MSC_VER) && !defined(PLATFORM_AMD64) && !defined(__clang__)
     const bool useAVX3MaskCmp = false;
-#else
+# else
     const bool useAVX3MaskCmp = (use_isa >= ISA_LEVEL_AVX3);
+# endif
 #endif
 
     __m128i lfCompare = _mm_set1_epi8('\n');

--- a/src/yencode/encoder_avx_base.h
+++ b/src/yencode/encoder_avx_base.h
@@ -294,7 +294,7 @@ HEDLEY_ALWAYS_INLINE void do_encode_avx2(int line_size, int* colOffset, const ui
                     asm(
                         "shrq $1, %[eqMask] \n"
                         "shrq %%cl, %[eqMask] \n"
-                        "adcq %[col], %[p] \n"
+                        "adcq %q[col], %q[p] \n"
                         : [eqMask]"+r"(eqMask), [p]"+r"(p)
                         : "c"(shiftAmt), [col]"r"(~col)
                     );

--- a/src/yencode/encoder_sse_base.h
+++ b/src/yencode/encoder_sse_base.h
@@ -424,8 +424,8 @@ do_encode_sse(int line_size, int *colOffset, const uint8_t *HEDLEY_RESTRICT srcE
                         asm(
                             "shrl $1, %[eqMask] \n"
                             "shrl %%cl, %[eqMask] \n" // TODO: can use shrq to avoid above shift?
-# if defined(PLATFORM_AMD64)
-                            "adcq %[col], %[p] \n"
+# if defined(PLATFORM_AMD64) && !defined(__ILP32__)
+                            "adcq %q[col], %q[p] \n"
 # else
                             "adcl %[col], %[p] \n"
 # endif

